Add documentation on writing templates (#395)
matrix-org/sydent: docs/templates.md

# Templates

Sydent uses parametrised templates to generate the content of emails it
sends and webpages it shows to users. Example templates can be found in
the [res](https://github.com/matrix-org/sydent/tree/main/res) folder.

## Branding

Sending a value for `brand` to some API endpoints allows for different
email and http templates to be used. These templates should be stored
in a file structure like this:

```
 root_template_dir/
     brand1/
         [ brand1 template files ]
     brand2/
         [ brand2 template files ]
```

The config option `templates.root_directory` should be set to the path
of `root_template_dir` and `templates.default_brand` should be set to
the sub directory to use if no `brand` (or an invalid `brand`) is provided
by the request.

Here are the requests that can contain a value for `brand` and what
template the brand value is used to select:

Invite email templates:

>  - `POST /_matrix/identity/v2/validate/msisdn/requestToken`

Author: Oops a copy paste error!

Verification email templates:
 - `POST /_matrix/identity/v2/validate/email/requestToken`

Verification SMS templates - BRAND CURRENTLY IGNORED:
 - `POST /_matrix/identity/v2/validate/msisdn/requestToken`

Verification response templates:
 - `GET /_matrix/identity/v2/validate/email/submitToken`
 - `GET /_matrix/identity/v2/validate/msisdn/submitToken`

## Template formats

### `.eml` template files

Files ending in ".eml" can use python `%(variable)s` string substitution.

Appending "_forurl" or "_forhtml" to any of the variable names listed below
returns their values encoded suitably for URLs or HTML respectively.

> e.g. ">" in the HTML encoded version would be replaced with "&gt".

Reviewer: I'm not even sure if we should be documenting how to do this. We should really encourage people to use the Jinja templates.

Author: Should I put a deprecated warning around them or just get rid of them?

If someone was trying to migrate from the `.eml` files to `.eml.j2` ones then it might be helpful to have documentation on how they work

Reviewer: Yeah, let's use this as an opportunity to officially deprecate them. (We'll need to ensure this goes in the changelog too.)

Reviewer 2: Personally I feel we don't need to document the old-school templates, we're hoping to see them gone soon

Author: I've put a deprecated notice next to the .eml information and updated the changelog, we can always remove the .eml documentation when we drop support?

> ### `.eml.j2` files

Reviewer: Per some other conversation, we might want to include that these need to be `\r\n` line-endings?

Author: Is that in both the jinja and the .eml files?

Reviewer 2: > these need to be \r\n line-endings?

oh god really. We ought to fix that in Sydent itself if we can...

Reviewer: I don't think sydent should be expecting raw eml files, but that's a much larger issue.

Author: > Per some other conversation, we might want to include that these need to be `\r\n` line-endings?

That shouldn't be an issue... The libary Sydent uses says that "lone \r and \n characters are converted to \r\n characters" https://github.com/python/cpython/blob/1afc7b3219b24c951bb4e6b7e1ead904228de074/Lib/smtplib.py#L808

Reviewer: This was fixed locally for the copies in ansible, I have not verified that they need to be `\r\n`, but there are comments referring to that.

Reviewer: (This might not be true 🤷 )

Files ending in ".eml.j2" are Jinja templates. 

Using the `urlencode` Jinja filter encodes the contents for URLs suitably.

> All variables are automatically made safe for HTML.

Reviewer: We probably want to include that some of them should use `safe` if they're not in HTML.

Author: What are the advantages of marking them as safe? Is there a significant performance boost?

Reviewer 2: It means things won't be double-escaped e.g. `>` → `&gt;` → `&amp;gt;` (renders as '&gt;').

Author: Where is the second layer of escaping happening?

Reviewer 2: >Where is the second layer of escaping happening?

my mistake, have just had a very cursed history with PHP and forgot we're talking about the plaintext section of the e-mail here

Files ending in ".eml" can use python `%(variable)s` string substitution.

Appending "_forurl" or "_forhtml" to any of the variable names listed below
returns their values encoded suitably for URLs or HTML respectively.

> For example ">" in the HTML encoded version would be replaced with "&gt".

Reviewer: I'm not sure what this line is referring to?

It might be worth mentioning that you need to double & to get a raw one. E.g. `&&` -> `&` after processing.

Author: What do you mean by referring to? 

By "listed below" I meant the tables of variables and line 65 was referring to the "_forhtml" suffix mentiond on line 62 if that answers your question?

Reviewer: Maybe this is clearer?

```suggestion
For example ">" in the `_forhtml` version would be replaced with "&gt".
```

---

## Invite email templates 

Invitation emails are sent when someone is invited to a room by their
email address. They should contain a link for the user to click on
that takes them to a matrix client.

Invite template files should have the name `invite_template.eml.j2`.

### Substitutions from Sydent

Variable                | Contents 
-----------             | -------- 
`date`                  | The time and date of sending as defined in RFC 2822 (e.g. "Fri, 09 Nov 2001 01:08:47 -0000") 
`ephemeral_private_key` | The ephemeral private key being used for this invite
`from`                  | The sending email address as configured in `email.from`
`messageid`             | The unique ID for this email 

> `multipart_boundary`    | Randomized multipart boundary to use in multipart emails

Reviewer: Note that this doesn't have a `_forurl` or `_forhtml` substitute. It might be worth moving this out of the individual email sections and saying it is available for all emails?

Author: I've added a comment next to it saying that _forurl and _forhtml formats aren't available. I'm hesitant to put it somewhere else as I think it might be use ful to have all of the variables for each template grouped together (and so putting just one off on its own might be overlooked)

`sender_avatar_url`     | The URL of the inviter's avatar
`sender_display_name`   | The current display name of the inviter

Version when table last updated: Synapse v1.42.0

## Verification email templates

Verification template files should have the name `verification_template.eml.j2`.

### All substitutions

Variable                | Contents 
-----------             | -------- 
`date`                  | The time and date of sending as defined in RFC 2822 (e.g. "Fri, 09 Nov 2001 01:08:47 -0000") 
`from`                  | The sending email address as configured in `email.from`
`ipaddress`             | The IP address that the verification request came from. If the IP address is unknown then this is the string "an unknown location"
`messageid`             | The unique ID for this email 
`multipart_boundary`    | Randomized multipart boundary to use in multipart emails
`to`                    | The destination email address 
`token`                 | A randomly generated token that some clients might need

## Verification SMS templates

The SMS template isn't read from a file. Instead the SMS template
should be placed as a string into the config option `bodyTemplate`.

> This string uses python `{variable}` substitution.

Reviewer: Meaning we call `.format(token=token)` on it?

Author: Yep!

### All substitutions

Variable    | Contents 
----------- | -------- 
`token`     | A randomly generated token for the user to enter into the client

## Migration email templates

> Migration template files should have the name `migration_template.eml.j2`.

Reviewer: It might be worth double mentioning that this must use jinja, there's no `.eml` version.

### All substitutions

Variable                | Contents 
-----------             | -------- 
`date`                  | The time and date of sending as defined in RFC 2822 (e.g. "Fri, 09 Nov 2001 01:08:47 -0000") 
`from`                  | The sending email address as configured in `email.from`
`messageid`             | The unique ID for this email 
`multipart_boundary`    | Randomized multipart boundary to use in multipart emails
`mxid`                  | The user ID that has been disassociated from the destination email address
`to`                    | The destination email address 

> ## Verification response http templates

Reviewer: Was this done?

> ## Verification response http templates

Author: Done now (#406)

 - `POST /_matrix/identity/v2/validate/msisdn/requestToken`

Verification response templates:
 - `GET /_matrix/identity/v1/validate/email/submitToken`
 - `GET /_matrix/identity/v2/validate/email/submitToken`

 - `GET /_matrix/identity/v1/validate/msisdn/submitToken`
 - `GET /_matrix/identity/v2/validate/msisdn/submitToken`

## Template formats

### `.eml.j2` files

Files ending in ".eml.j2" are Jinja templates. 

Using the `urlencode` Jinja filter encodes the contents for URLs suitably.
All variables are automatically made safe for HTML.

If needed the `safe` filter can be used to prevent the HTML encoding, this is
useful for email headers and for including a plaintext portion of an email.

See the Jinja [documentation](https://jinja.palletsprojects.com/en/3.0.x/templates/#variables)
for more instructions on how to write these templates.

### DEPRECATED `.eml` template files 

Files ending in ".eml" can use python `%(variable)s` string substitution.

Appending "_forurl" or "_forhtml" to any of the variable names listed below
returns their values encoded suitably for URLs or HTML respectively.

For example ">" in the HTML encoded version would be replaced with "&gt".

> Note: "&&" when HTML encoded is displayed as "&".

Reviewer: ```suggestion
Note: "&&" must be used to get a raw `&` character.
```

> Note: Note: "&&" must be used to get a raw `&` character.

Reviewer: ```suggestion
Note: "&&" must be used to get a raw `&` character.
```
